src/api/MainSolver.cc

```diff
@@ -37,124 +19,182 @@ WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+                if (logic.isBooleanOperator(fla)) {
+                    PTRef old = fla;
+                    fla = rewriteMaxArity(fla);
+                    pmanager.transferPartitionMembership(old, fla);
+                }
+                assert(pmanager.getPartitionIndex(fla) != -1);
+                pmanager.propagatePartitionMask(fla);
+                status = giveToSolver(fla, frames[i].getId());
+                if (status == s_False) { break; }
+            }
+        } else {
+            PTRef frameFormula = logic.mkAnd(frames[i].formulas);
+            if (context.frameCount > 0) {
+                frameFormula = applyLearntSubstitutions(frameFormula);
+            }
+            frameFormula = theory->preprocessBeforeSubstitutions(frameFormula, context);
+            frameFormula = substitutionPass(frameFormula, context);
+            frameFormula = theory->preprocessAfterSubstitutions(frameFormula, context);
+
+            if (logic.isFalse(frameFormula)) {
+                giveToSolver(logic.getTerm_false(), frames[i].getId());
+                status = s_False;
+                break;
+            }
+            // Optimize the dag for cnfization
+            if (logic.isBooleanOperator(frameFormula)) {
+                frameFormula = rewriteMaxArity(frameFormula);
+            }
+            status = giveToSolver(frameFormula, frames[i].getId());
+        }
+    }
+    if (status == s_False) {
+        assert(firstNotSimplifiedFrame > 0);
+        rememberUnsatFrame(firstNotSimplifiedFrame - 1);
+    }
+    return status;
+}
+
+MainSolver::MainSolver(Logic& logic, SMTConfig& conf, std::string name)
+    :
+      theory(createTheory(logic, conf)),
+      term_mapper(new TermMapper(logic)),
+      thandler(new THandler(getTheory(), *term_mapper)),
+      smt_solver(createInnerSolver(conf, *thandler)),
+      logic(thandler->getLogic()),
+      pmanager(logic),
+      config(conf),
+      ts(logic, *term_mapper),
+      solver_name {std::move(name)}
+{
+    conf.setUsedForInitiliazation();
+    initialize();
+}
+
+MainSolver::MainSolver(std::unique_ptr<Theory> th, std::unique_ptr<TermMapper> tm, std::unique_ptr<THandler> thd,
+           std::unique_ptr<SimpSMTSolver> ss, Logic & logic, SMTConfig & conf, std::string name)
+    :
+      theory(std::move(th)),
+      term_mapper(std::move(tm)),
+      thandler(std::move(thd)),
+      smt_solver(std::move(ss)),
+      logic(thandler->getLogic()),
+      pmanager(logic),
+      config(conf),
+      ts(logic,*term_mapper),
+      solver_name {std::move(name)}
+{
+    conf.setUsedForInitiliazation();
+    initialize();
+}
+
+void MainSolver::initialize() {
+    frames.push();
+    frameTerms.push(logic.getTerm_true());
+    substitutions.push();
+    smt_solver->initialize();
+    opensmt::pair<CRef, CRef> iorefs{CRef_Undef, CRef_Undef};
+    smt_solver->addOriginalSMTClause({term_mapper->getOrCreateLit(logic.getTerm_true())}, iorefs);
```

Comment:
This corresponds to the code that was in `Cnfizer::initializer()` before.
I am not sure if we could go without it, maybe we could, but I have not tried.
It is just asserting to the underlying SAT solver, that literals corresponding to `true` and `not false` should be assigned to true (because we are adding unit clauses here).

Reply:
Ok, sounds fair, if it works let's not change it (for now at least)!